src/core/server/saved_objects/service/lib/collect_multi_namespace_references.ts

```diff
@@ -0,0 +1,306 @@
+ *
+ * @public
+ */
+export interface SavedObjectsCollectMultiNamespaceReferencesResponse {
+  objects: SavedObjectReferenceWithContext[];
+}
+
+export interface CollectMultiNamespaceReferencesParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: RepositoryEsClient;
+  serializer: SavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  objects: SavedObjectsCollectMultiNamespaceReferencesObject[];
+  options?: SavedObjectsCollectMultiNamespaceReferencesOptions;
+}
+
+/**
+ * Additional options exposed for unit tests.
+ *
+ * @internal
+ */
+export interface CollectMultiNamespaceReferencesParamsInternal
+  extends CollectMultiNamespaceReferencesParams {
+  /** Default: 1000 */
+  searchPerPage?: number;
+}
+
+/**
+ * Gets all references and transitive references of the given objects. Ignores any object and/or reference that is not a multi-namespace
+ * type.
+ */
+export async function collectMultiNamespaceReferences({
+  registry,
+  allowedTypes,
+  client,
+  serializer,
+  getIndexForType,
+  objects,
+  options = {},
+  searchPerPage,
+}: CollectMultiNamespaceReferencesParamsInternal): Promise<SavedObjectsCollectMultiNamespaceReferencesResponse> {
+  if (!objects.length) {
+    return { objects: [] };
+  }
+  const { namespace, purpose, typesToExclude = [] } = options;
+  const inboundReferencesMap = objects.reduce(
+    // Add the input objects to the references map so they are returned with the results, even if they have no inbound references
+    (acc, { type, id }) => acc.set(`${type}:${id}`, new Map()),
+    new Map<string, Map<string, string>>()
+  );
+  const objectMap = new Map<string, SavedObject | null>();
+
+  const rootFields = getRootFields();
+  const makeBulkGetDocs = (objectsToGet: SavedObjectsCollectMultiNamespaceReferencesObject[]) =>
+    objectsToGet.map(({ type, id }) => ({
+      _id: serializer.generateRawId(undefined, type, id),
+      _index: getIndexForType(type),
+      _source: rootFields, // Optimized to only retrieve root fields (ignoring type-specific fields)
+    }));
+  const objectFilter = ({ type }: SavedObjectsCollectMultiNamespaceReferencesObject) =>
+    allowedTypes.includes(type) &&
+    (purpose === 'updateObjectsSpaces'
+      ? registry.isShareable(type)
+      : registry.isMultiNamespace(type)) &&
+    !typesToExclude.includes(type);
+
+  let bulkGetObjects = objects.filter(objectFilter);
+  let count = 0; // this is a circuit-breaker to ensure we don't hog too many resources; we should never have an object graph this deep
+  while (bulkGetObjects.length && count++ < 20) {
+    const bulkGetResponse = await client.mget(
+      { body: { docs: makeBulkGetDocs(bulkGetObjects) } },
+      { ignore: [404] }
+    );
+    const newObjectsToGet = new Set<string>();
+    for (let i = 0; i < bulkGetObjects.length; i++) {
+      const { type, id } = bulkGetObjects[i];
+      const objectKey = `${type}:${id}`;
+      const doc = bulkGetResponse.body.docs[i];
```

Comment:
The link between the index in the `bulkGetResponse.body.docs` array and the index in the `bulkGetObjects` array is not super obvious to me. But I assume that the two are tightly corrected somehow?

Reply:
From my understanding, we can be certain that the ES `mget` response will contain a doc object (found or not) for each requested document, in the same order.